Update documentation organization, fix some layout issues, add BIDSPath introductory example

hoechenberger opened the pull request:
PR Description
--------------

Closes #526 
Closes #584

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [ ] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"

hoechenberger commented:
Doc artifact:
https://3251-89170358-gh.circle-artifacts.com/0/dev/index.html

I've touched ~everything documentation-related, so I'd appreciate your feedback.

hoechenberger commented:
cc @adam2392

adam2392 commented on examples/bidspath.py:
```diff
@@ -0,0 +1,160 @@
+
+bids_path.basename
+
+# The two
```
Combine with line below?

hoechenberger replied:
Done in f4ad8a6

sappelhoff commented on setup.py:
```diff
@@ -35,7 +35,7 @@
           download_url=DOWNLOAD_URL,
-          long_description=open('README.rst').read(),
+          long_description=open('README.md').read(),
           long_description_content_type='text/x-rst',
```
this would have to be adjusted for markdown, or does text/x-rst now work for both markdown and rst?

hoechenberger replied:
Whoops. Yes probably needs adjustment.

hoechenberger replied:
Done in 57c8733

sappelhoff commented on doc/whats_new.rst:
```diff
@@ -7,312 +7,139 @@
+- `Alexandre Gramfort`_
+- `Alex Rockhill`_
+- `Richard Höchenberger`_ 
+- `Stefan Appelhoff`_
```
need to add Mainak, maybe you forgot him because he didn't make a commit for this release?

hoechenberger replied:
Right, I only included those who committed, so I forgot about Mainak's extensive code reviews! Will add.

hoechenberger replied:
Resolved in ef4da08

jasmainak commented on setup.py:
```diff
@@ -35,8 +35,8 @@
           download_url=DOWNLOAD_URL,
-          long_description=open('README.rst').read(),
-          long_description_content_type='text/x-rst',
+          long_description=open('README.md').read(),
```
what's the reason for this?

hoechenberger replied:
I reworked the readme and other parts of the documentation entirely, and GitHub markdown is just easier to read & write than reST. That's the only reason really.

jasmainak commented on examples/bidspath.py:
```diff
@@ -0,0 +1,159 @@
+"""
+=========================================
+An introduction to BIDSPath
+=========================================
```
we normally don't make the === go longer than the text. Just a stylistic nitpick :)

hoechenberger replied:
I've changed the title a couple times and apparently overlooked this. will adjust.

hoechenberger replied:
Done in 879eaf5

jasmainak commented on examples/bidspath.py:
```diff
@@ -0,0 +1,158 @@
+"""
+===========================
+An introduction to BIDSPath
```
aren't the examples ordered in terms of difficulty?

hoechenberger replied:
I wasn't sure where to put this one, so I simply put it at the very end. Any suggestions on how to make it better are of course appreciated!

jasmainak replied:
do we still need 01. with this tutorial? Are they duplicating?

hoechenberger replied:
> Are they duplicating?

I'm not sure what you mean by that..?

sappelhoff replied:
I think @jasmainak's point here is to scrap our current 01. example ("Creating BIDS-compatible folder names and filenames") and expand the bidspath example to cover everything from that old 01. example

--> and then make the bidspath example the new 01. example, which would be fitting, because it's now becoming the center piece of mne-bids.

I would be in favor of that.

hoechenberger replied:
I will give this a shot

jasmainak replied:
Thanks for expanding the thoughts in my mind @sappelhoff :)

sappelhoff commented on doc/_templates/navbar.html:
```diff
@@ -9,6 +9,7 @@
   <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
     <li><a href="https://mne.tools/mne-bids/dev/index.html">Development</a></li>
     <li><a href="https://mne.tools/mne-bids/stable/index.html">Stable</a></li>
+    <li><a href="https://mne.tools/mne-bids/v0.5/index.html">v0.5</a></li>
```
just a note: this will lead to a 404 until the actual release is done, docs are built, and manually pushed to the gh-pages branch (see wiki "how to make a release")

sappelhoff commented on doc/authors.rst:
```diff
@@ -0,0 +1,19 @@
+.. _Sophie Herbst: http://github.com/SophieHerbst
+.. _Adam Li: https://github.com/adam2392
+.. _Fu-Te Wong: https://github.com/zuxfoucault
+.. _Richard Höchenberger: https://github.com/hoechenberger
```
why is your name gray here @hoechenberger or is it only in my UI?

![grafik](https://user-images.githubusercontent.com/9084751/96504416-f0b6d980-1254-11eb-8c7b-51e53aacd7b6.png)

hoechenberger replied:
I have no idea! Hopefully it's a good thing LOL

sappelhoff commented on doc/conf.py:
```diff
@@ -76,7 +76,8 @@
 # directories to ignore when looking for source files.
 # This patterns also effect to html_static_path and html_extra_path
-exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
+exclude_patterns = ['auto_examples/index.rst', '_build', 'Thumbs.db',
```
why do we have to exclude the index.tst from auto_examples?

hoechenberger replied:
Because otherwise Sphinx would try to process it automatically, and since we're manually `include`ing it in `use.rst` again, it would throw a warning about the same ID used twice. Adding it to the `exclude_patterns` seemed like the easiest way for me to fix it, but I'm not a Sphinx expert…

sappelhoff replied:
okay, that sounds reasonable. thanks.

sappelhoff commented on doc/use.rst:
```diff
@@ -0,0 +1,59 @@
+
+Quickstart
+----------
+MNE-BIDS fully supports writing of BIDS datasets for MEG and EEG. Support for
+iEEG is experimental at the moment.
```
is iEEG support still experimental? I feel like there have been many steps towards full ieeg functionality by @alexrockhill @adam2392 et al

hoechenberger replied:
I've removed this line in 0a5af78

sappelhoff commented on doc/conf.py:
```diff
@@ -101,11 +102,12 @@
+        ("Install", "install"),
+        ("Use", "use"),
+        # ("API", "api"),
+        # ("CLI", "generated/cli"),
```
the CLI docs do not seem to work: https://3251-89170358-gh.circle-artifacts.com/0/dev/generated/cli  --> clicked on a link here:
![grafik](https://user-images.githubusercontent.com/9084751/96505043-e0532e80-1255-11eb-8391-e96974914f39.png)

hoechenberger replied:
Thank you, this is now fixed via 41c1cfd

agramfort commented on setup.py:
```diff
@@ -35,8 +35,8 @@
-          long_description=open('README.rst').read(),
-          long_description_content_type='text/x-rst',
+          long_description=open('README.md').read(),
+          long_description_content_type='text/markdown',
```
I don't understand why this was changed but whatever works

sappelhoff replied:
@hoechenberger's argument was that markdown is easier for him.

That's okay for me as long as we carefully check how this renders on https://pypi.org/project/mne-bids/ before releasing (by making use of test.pypi).

jasmainak replied:
one counterargument is that rst allows you to build in sphinx-gallery. Actually, I am pretty sure that was the reason I kept it in rst initially

agramfort replied:
> @hoechenberger's argument was that markdown is easier for him.

hum... so next time I touch I switch bck to rst cause it's easier for me :)

sappelhoff replied:
> one counterargument is that rst allows you to build in sphinx-gallery without duplicating. Actually, I am pretty sure that was the reason I kept it in rst initially

yes, that was the reason (which is still valid).

hoechenberger replied:
Oh my what a discussion!

So initially, `README.rst` would be what users would see when visiting the GitHub repo, and the exact same file would be included as the landing page of our documentation by Sphinx.

I totally reworked this, essentially moving the previous `README.rst` to `doc/index.rst`, and changing some things; and basically creating a new (and massively reduced) `README.md` based on a few blocks of existing text, but otherwise basically from scratch. So to me it was like creating a new file, not changing an old one, because the old one was moved.

If people feel so strongly about using reST instead of markdown, I can of course convert it to reST again

hoechenberger replied:
So to summarize:

`README.md` is not included in the built docs, it's really just for those who come and stop by to look at the GH repo. The old (and updated) `README.rst` is now `doc/index.rst`.

agramfort replied:
do not revert now that it's done. All ok on my side

sappelhoff replied:
also ok from my side, I actually like a short and to-the-point GitHub README and a longer and explicit documentation index.rst

jasmainak replied:
yeah it's fine for me too :)

jasmainak commented on doc/api.rst:
```diff
@@ -39,28 +35,12 @@ MNE BIDS
-   :toctree: generated/
-
-   BIDSPath
-   get_entities_from_fname
```
why are these gone now?

hoechenberger replied:
Because they're exposed from `mne_bids/__init__.py` (and this is how we import them in all our examples: `from mne_bids import BIDSPath`)

They're still listed in the API docs – just under `mne_bids`, not `mne_bids.path` anymore :)

jasmainak commented:
Just peeked at the redesign, looks lovely!

jasmainak commented on doc/api.rst:
```diff
@@ -6,16 +6,12 @@
 API Documentation
 =================
```
so I noticed that this page is gone from the navbar. It's of course subjective but the API documentation is something I look at *pretty* often for any project. Having to do two clicks to get there is not the most optimal I would say. My suggestion would be to add it back between "Use" and "Contribute"

alexrockhill replied:
+1

hoechenberger replied:
I've added it back in a786be5

hoechenberger replied:
@jasmainak @alexrockhill Let me know if the current solution is good for you:

<img width="708" alt="Screenshot 2020-10-21 at 09 46 13" src="https://user-images.githubusercontent.com/2046265/96689113-450ba780-1382-11eb-8aee-7d4bebc8ce06.png">


https://3262-89170358-gh.circle-artifacts.com/0/dev/index.html

jasmainak commented on doc/contribute.rst:
```diff
@@ -0,0 +1,29 @@
+Development takes place on the collaborative platform GitHub at
+`github.com/mne-tools/mne-bids <https://github.com/mne-tools/mne-bids>`_.
+
+.. image:: assets/GitHub.png
```
how about adding it to the gh-pages branch directly to avoid bloating master?

hoechenberger replied:
It's 230 kB, does it really matter?

sappelhoff replied:
I gave Mainak's comment a thumbs up, but reconsidering it now --> it's probably cleaner to have the documentation source + all assets in one place. So +1 to have it like it is here. The files are small enough for my taste.

jasmainak replied:
you can't remove a thumbs up :) I think 230 kB will matter if ever you want to update that image. You won't be able to force push. So if you're certain you'll never change it, then go for it.

hoechenberger replied:
Maybe the compromise is to include the files in this repo for now, and if we realize during future doc updates that we keep on changing this stuff, we move the images to gh-pages. What do you think?





> you can't remove a thumbs up :)

Yeah there should be an edit history haha :)

jasmainak commented on examples/bidspath.py:
```diff
@@ -0,0 +1,159 @@
+run = '01'
+session = None
+bids_path.update(run=run, session=session)
+bids_path
```
print(bids_path) ?

to avoid confusing the user with two different ways to do similar things

hoechenberger replied:
The thing is that `print(bids_path)` produces a different output than just `bids_path`, because `BIDSPath.__str__` and `BIDSPath.__repr__` are different. But you're right, it might be confusing

adam2392 replied:
the `str` essentially calls `fpath`, whereas `repr` is more like Pathlib looking. IIRC I think functionally this made it a lot easier to pass around bids_path to functions that might require a string (e.g. some `os` functions).

I think that's what people defaulted to when reviewing the initial BIDSPath.

hoechenberger replied:
Yes I remember this discussion too. Just to recap for all of you, this is what it currently looks like:

```python
In [5]: p = BIDSPath(root='/tmp/foo', subject='01', session='test',
   ...:              task='dostuff', datatype='meg')

In [6]: p
Out[6]:
BIDSPath(
root: /tmp/foo
datatype: meg
basename: sub-01_ses-test_task-dostuff)

In [7]: print(p)
/Users/hoechenberger/Development/mne-bids/mne_bids/path.py:445: RuntimeWarning: Could not locate a data file of a supported format. This is likely a problem with your BIDS dataset. Please run the BIDS validator on your data. (root=/tmp/foo, basename=sub-01_ses-test_task-dostuff). []
  warn(msg)
/tmp/foo/sub-01/ses-test/meg/sub-01_ses-test_task-dostuff
```
I'm not so sure about this warning either…
**Plus** the above lines create a new folder structure `/tmp/foo/sub-01/ses-test/meg`, which is probably not what we want either? (`/tmp/foo` did not exist before)

cc @agramfort

hoechenberger replied:
No warning when I specify suffix and extension:

```python
In [14]: p = BIDSPath(root='/tmp/foo', subject='01', session='test',
    ...:              task='dostuff', datatype='meg', suffix='meg',
    ...:              extension='.fif')

In [15]: print(p)
/tmp/foo/sub-01/ses-test/meg/sub-01_ses-test_task-dostuff_meg.fif
```

hoechenberger replied:
But generally I like the `repr`. And the `str` being path-like makes it easier to pass a `BIDSPath` to other functions, as @adam2392 pointed out. I will see if and how I can adjust the example without us having to change `BIDSPath`.

agramfort replied:
yes str is to be able to pass a BIDSPath as a regular path in our function. It makes it's path-like. The repr is just for user facing information so it's ok that repr and str are different.

sappelhoff replied:
okay, makes sense!

hoechenberger replied:
I'm addressing this in b50b260, check out example 01.

jasmainak commented on examples/bidspath.py:
```diff
@@ -0,0 +1,159 @@
+###############################################################################
+# By default, most MNE-BIDS functions will try to infer to correct
+# suffix and extension for your data, and you don't need to specify them
+# manually.
```
isn't there some functionality to query as well? cc @adam2392 should perhaps review this example too

hoechenberger replied:
Yes we added `BIDSPath.match` for example, is that what you mean?

jasmainak replied:
yes!

adam2392 replied:
The example overall looks pretty good to me. It makes sense to merge the 01-example based on what I've seen there. You all critiqued thoroughly already :0

jasmainak commented on doc/whats_new.rst:
```diff
@@ -7,312 +7,140 @@
+- `Adam Li`_
+- `Alexandre Gramfort`_
+- `Alex Rockhill`_
+- `Mainak Jas`_
```
thanks for adding me :)

I see a name missing: Evgenii Kalenkovich

In MNE, we typically look at the commits. You can find a command in the release documentation. That way you don't miss out anyone.

alexrockhill replied:
Austin Hurst (#456) maybe also missing

sappelhoff replied:
```
 git shortlog v0.4..HEAD --summary --numbered --email
    41  Richard Höchenberger <richard.hoechenberger@gmail.com>
    35  Adam Li <adam2392@gmail.com>
    16  Alexandre Gramfort <alexandre.gramfort@m4x.org>
    13  Stefan Appelhoff <stefan.appelhoff@mailbox.org>
    10  kalenkovich <e.kalenkovich@gmail.com>
     6  Austin Hurst <mynameisaustinhurst@gmail.com>
     4  Alex Rockhill <aprockhill@mailbox.org>
     2  Alex Rockhill <aprockhill206@gmail.com>
     1  Robert Luke <748691+rob-luke@users.noreply.github.com>
```

we also need to update [.mailmap](https://github.com/mne-tools/mne-bids/blob/master/.mailmap) to map yet another email of Alex R :rofl:  --> and to maybe get the true name of kalenkovich and properly map it.

hoechenberger replied:
Resolved in 4700e1a

sappelhoff replied:
lgtm, thanks

jasmainak commented on examples/bidspath.py:
```diff
@@ -0,0 +1,159 @@
+# ``sub``, this is ``123``, and for ``ses``, it is ``test`` in our example.
+# Entity names (or "keys") and values are separated via hyphens.
+# BIDS knows a much larger number of entities, and MNE-BIDS allows you to make
+# use of them. To get a list of all supported entities, use:
```
```suggestion
# use of them. To get a list of all entities in bids_path, use:
```

I presume you don't get a list of all entities supported by BIDS?

hoechenberger replied:
You get a list of all entities supported in MNE-BIDS

hoechenberger replied:
@jasmainak 

```python
In [25]: p = BIDSPath(root='/tmp/foo', subject='01', session='test',
    ...:              task='dostuff', datatype='meg', suffix='meg',
    ...:              extension='.fif')

In [26]: p.entities
Out[26]:
OrderedDict([('subject', '01'),
             ('session', 'test'),
             ('task', 'dostuff'),
             ('acquisition', None),
             ('run', None),
             ('processing', None),
             ('space', None),
             ('recording', None),
             ('split', None)])
```

sappelhoff reviewed (commented):
just wanted to say thanks @hoechenberger this looks good to me overall and I am +1 for merging and moving on to a release once all comments in this PR are resolved.

agramfort commented:
@sappelhoff @jasmainak I let merge when you're happy

hoechenberger commented:
I've reworked the examples:

- the first example now demonstrated how to read BIDS data, and explores a bit (but not to full extent) how to use `BIDSPath`
- the second example is now writing BIDS data (conversion MNE sample --> BIDS)
- the previously first example, "conversion of BIDS-compatible folder and file names", is now placed after all other examples, along with the `BIDSPath` introduction.

I think this makes much more sense now because the first examples should now help get you started immediately, without being introduced to BIDS folder structures etc which you may not even need to fully understand if you just want to open a dataset.

hoechenberger commented:
Build artifact:
https://3264-89170358-gh.circle-artifacts.com/0/dev/use.html#examples

sappelhoff reviewed (changes requested):
new structure looks good to me! Could you add the zero-padded numbers to the last two examples for consistency, please? 01-example1, ... 09-example9, ...

hoechenberger commented:
> new structure looks good to me! Could you add the zero-padded numbers to the last two examples for consistency, please? 01-example1, ... 09-example9, ...

In fact, I was considering to remove those numbers altogether, as I feel they are creating the feel of a stronger hierarchy than there actually is… thoughts?

sappelhoff commented:
> In fact, I was considering to remove those numbers altogether, as I feel they are creating the feel of a stronger hierarchy than there actually is… thoughts?

they were added to properly order the examples (i.e., make them appear in a sensible order and from general to specific in terms of content)

if you can manage to get such an order without the numbers, that'd be awesome.

hoechenberger commented:
> if you can manage to get such an order without the numbers, that'd be awesome.

Seems this is not trivial, so I just added the numbers :)

I believe I have addressed all comments now, except for the question on where to put the images. I don't think that we've reached a consensus there? I'm not in the mood to discuss, so consider my vote as "neutral" ;) and I'll just do whatever you think is best.

hoechenberger commented:
Latest documentation artifact:
https://3265-89170358-gh.circle-artifacts.com/0/dev/index.html

sappelhoff reviewed (approved):
>  I'm not in the mood to discuss, so consider my vote as "neutral"

:)

I think the changes in this PR are great, +1 to merge.

agramfort commented:
+1 for MRG once CIs are fixed https://github.com/mne-tools/mne-bids/pull/588/checks?check_run_id=1288539086

jasmainak commented on doc/use.rst:
```diff
@@ -0,0 +1,60 @@
+   :local:
+   :depth: 1
+
+Python API reference
+--------------------
+See the :doc:`API documentation <api>`.
+
+Command line interface reference
+--------------------------------
+See the :doc:`CLI documentation <generated/cli>`.
```
this is now superfluous?

jasmainak commented on examples/read_bids_datasets.py:
````diff
@@ -1,82 +1,172 @@
+# ------------------------
+#
+# There is obviously only one subject (``01``) and one experimental task
+# (``somato```). Let's use this knowledge to create a new ``BIDSPath`` with
````
triple backticks

jasmainak commented on examples/read_bids_datasets.py:
```diff
@@ -1,82 +1,172 @@
-# Step 2: Read a BIDS dataset
-# ---------------------------
+# But what about that filename extension, now? ``BIDSPath.fpath``, which –
+# as you hopefully remember – is invoked when you to ``print(bids_path)`` –
```
when you to -> when you do

jasmainak commented on examples/read_bids_datasets.py:
```diff
@@ -1,82 +1,172 @@
+# employs some heuristics to auto-detect some missing filename components.
+# This can come in handy if, for example, you intend to re-use the same
+# MNE-BIDS script on data acquired with different acquisition systems, and
+# accordinly stored in different formats and with different filename
```
accordinly -> accordingly

jasmainak commented on examples/read_bids_datasets.py:
```diff
@@ -1,82 +1,172 @@
+# But what about that filename extension, now? ``BIDSPath.fpath``, which –
+# as you hopefully remember – is invoked when you to ``print(bids_path)`` –
+# employs some heuristics to auto-detect some missing filename components.
+# This can come in handy if, for example, you intend to re-use the same
```
This feels like a very niche example that should be left out of the first example. I'd rather just "show and tell"

jasmainak replied:
I mean the example of using a "different acquisition system"

jasmainak commented on examples/read_bids_datasets.py:
```diff
@@ -1,82 +1,172 @@
+# is the same) into your Python console, you would get the nicely formatted
+# output:
+
+bids_path
```
you should tell the user what is "root" and "basename", no?

jasmainak commented on examples/mark_bad_channels.py:
```diff
@@ -1,6 +1,6 @@
 """
 ===============================================
-10. Changing which channels are marked as "bad"
+03. Changing which channels are marked as "bad"
```
I just noticed L28 in this example. That's bad, you're setting the MNE log level silently whenever user runs the example.

jasmainak reviewed (approved):
Looks great. Thanks for all the efforts @hoechenberger !

One last comment. I am wondering if we should have something in the examples for functions that are currently "dangling". Example: `get_entity_vals` and `get_datatypes` don't figure in any example.

hoechenberger commented:
I have addressed all review comments.

@jasmainak I'm not setting global MNE log level anymore in example 03. This, however, causes way too much distracting output. I'm okay with keeping it that way for now, but we need to reconsider logging in general. I believe @adam2392 also thought we're printing way too much noise by default.

I've now also pushed the images to `gh-pages` because otherwise it's not possible to display images in the readme on PyPI:
https://test.pypi.org/project/mne-bids/0.5.dev2/

jasmainak commented:
> This, however, causes way too much distracting output. I'm okay with keeping it that way for now, but we need to reconsider logging in general.

why not set the `verbose` parameter for that rather than setting the global log level?

hoechenberger commented:
> why not set the `verbose` parameter for that rather than setting the global log level?

I think I tried and it behaves differently. Will try again sometime
